File: email-payments.ts
// Email-based payment verification via IMAP.
// Supports Cash App, Venmo, Zelle, and PayPal notification emails.
// All providers share one IMAP inbox (one set of credentials per guild).

import { ImapFlow } from 'imapflow';

export type PaymentProvider = 'cashapp' | 'venmo' | 'zelle' | 'paypal';

export interface PaymentEmailConfig {
  email: string;
  password: string;
  host?: string;
  port?: number;
}

const FETCH_TIMEOUT_MS = 15000;

const KNOWN_HOSTS: Record<string, string> = {
  'gmail.com':      'imap.gmail.com',
  'googlemail.com': 'imap.gmail.com',
  'outlook.com':    'outlook.office365.com',
  'hotmail.com':    'outlook.office365.com',
  'live.com':       'outlook.office365.com',
  'yahoo.com':      'imap.mail.yahoo.com',
  'ymail.com':      'imap.mail.yahoo.com',
  'icloud.com':     'imap.mail.me.com',
  'me.com':         'imap.mail.me.com',
};

function inferHost(email: string): string {
  const domain = email.split('@')[1]?.toLowerCase() ?? '';
  return KNOWN_HOSTS[domain] ?? 'imap.gmail.com';
}

// ── Per-provider email detection patterns ─────────────────────────────────────

const PROVIDER_PATTERNS: Record<PaymentProvider, { from: RegExp[]; subject: RegExp[] }> = {
  cashapp: {
    from: [
      /cash@square\.com/i,
      /no-?reply@cash\.app/i,
      /@cashapp\.com/i,
      /cash@cashapp\.com/i,
    ],
    subject: [
      /sent you \$/i,
      /you received \$/i,
      /cash app.*payment/i,
      /payment.*cash app/i,
      /you've got cash/i,
    ],
  },
  venmo: {
    from: [
      /venmo@venmo\.com/i,
      /no-?reply@venmo\.com/i,
    ],
    subject: [
      /paid you/i,
      /sent you \$/i,
      /you've got money/i,
      /payment.*received/i,
      /venmo.*payment/i,
    ],
  },
  zelle: {
    from: [
      /zelle/i,
      /noreply@chase\.com/i,
      /alerts@bankofamerica\.com/i,
      /alerts@.*wellsfargo\.com/i,
      /alerts@.*usbank\.com/i,
      /alerts@.*capitalone\.com/i,
      /notify@.*zellepay\.com/i,
    ],
    subject: [
      /you received.*zelle/i,
      /received money with zelle/i,
      /zelle.*payment received/i,
      /payment.*received.*zelle/i,
      /you have received/i,
      /zelle.*you.*received/i,
    ],
  },
  paypal: {
    from: [
      /service@paypal\.com/i,
      /service@intl\.paypal\.com/i,
      /paypal@e\.paypal\.com/i,
      /noreply@paypal\.com/i,
    ],
    subject: [
      /you've got money/i,
      /you received \$/i,
      /payment.*received/i,
      /sent you \$/i,
      /paypal.*payment/i,
      /money.*received/i,
    ],
  },
};

function isProviderEmail(provider: PaymentProvider, subject: string, from: string): boolean {
  const p = PROVIDER_PATTERNS[provider];
  const fromMatch    = p.from.some(r => r.test(from));
  const subjectMatch = p.subject.some(r => r.test(subject));
  // Cash App: require BOTH correct sender AND a payment-received subject to exclude billing/overdue emails
  if (provider === 'cashapp') return fromMatch && subjectMatch;
  return fromMatch || subjectMatch;
}

function extractMemo(text: string): string {
  // Cash App email format: "For XMO4917H" — "For" followed by a space and the order ID on the same line.
  // Also handle "Note: <text>", "Memo: <text>" for other providers.
  // We intentionally reject long generic phrases and technical strings.
  const patterns: RegExp[] = [
    // Cash App: "For <order-id>" — short alphanumeric token, with or without colon
    /\bFor:?\s+([A-Z0-9][A-Z0-9\-_]{1,40})\b/i,
    // If the word "For" is on its own line followed by value on next line
    /\bFor\s*[\r\n]+\s*([A-Z0-9][A-Z0-9\-_]{1,40})\b/i,
    // Generic labeled fields
    /\bNote:\s*([^\n\r<]{2,100})/i,
    /\bNote\s*[\r\n]+\s*([^\n\r<]{2,100})/i,
    /\bMemo:\s*([^\n\r<]{2,100})/i,
    /\bMessage:\s*([^\n\r<]{2,100})/i,
    /\bDescription:\s*([^\n\r<]{2,100})/i,
    /\bReference:\s*([^\n\r<]{2,100})/i,
    // Venmo payment note: "paid you with note: ..."
    /note:?\s+"([^"]{2,100})"/i,
    /note:?\s+'([^']{2,100})'/i,
  ];
  for (const re of patterns) {
    const m = re.exec(text);
    if (!m) continue;
    const val = m[1].trim().replace(/\s+/g, ' ');
    if (val.length < 2) continue;
    // Reject technical/garbage strings
    if (/https?:\/\//i.test(val)) continue;
    if (/version=|cipher|bits\d|sha\d{3}/i.test(val)) continue;
    if (/[0-9a-f]{20,}/i.test(val)) continue;
    if (/@\w+\.\w+/.test(val)) continue;
    // Reject common generic phrases that aren't memos
    if (/^(more information|questions|your account|any questions|details|security|help|click|view|manage|update|visit)\b/i.test(val)) continue;
    return val;
  }
  return '';
}

function stripRFC822Headers(raw: string): string {
  // RFC822: blank line separates headers from body
  const crlf = raw.indexOf('\r\n\r\n');
  if (crlf !== -1) return raw.slice(crlf + 4);
  const lf = raw.indexOf('\n\n');
  if (lf !== -1) return raw.slice(lf + 2);
  return raw;
}

function decodeQuotedPrintable(str: string): string {
  return str
    .replace(/=\r?\n/g, '')
    .replace(/=([0-9A-Fa-f]{2})/g, (_, hex) => String.fromCharCode(parseInt(hex, 16)));
}

function extractPlainText(raw: string): string {
  return decodeHtmlEntities(
    decodeQuotedPrintable(stripRFC822Headers(raw))
      .replace(/<[^>]+>/g, ' ')
  );
}

function decodeHtmlEntities(str: string): string {
  return str
    .replace(/&amp;/g,   '&')
    .replace(/&nbsp;/g,  ' ')
    .replace(/&#160;/g,  ' ')
    .replace(/&lt;/g,    '<')
    .replace(/&gt;/g,    '>')
    .replace(/&quot;/g,  '"')
    .replace(/&#39;/g,   "'")
    .replace(/&#x27;/g,  "'")
    .replace(/&#x2F;/g,  '/')
    .replace(/&#(\d+);/g, (_, code) => String.fromCharCode(Number(code)));
}

function extractAmounts(text: string): number[] {
  const amounts: number[] = [];
  // Match $1.00, +$1.00, $1,234.50, $ 12.50 (Cash App uses + prefix for received payments)
  const re = /[+\-]?\$\s*([\d]{1,6}(?:,\d{3})*(?:\.\d{1,2})?)/gi;
  let m: RegExpExecArray | null;
  while ((m = re.exec(text)) !== null) {
    const v = parseFloat(m[1].replace(/,/g, ''));
    if (!isNaN(v) && v > 0 && v < 10000) amounts.push(v);
  }
  return [...new Set(amounts)]; // deduplicate
}

function makeClient(config: PaymentEmailConfig): ImapFlow {
  const host = config.host || inferHost(config.email);
  const port = config.port || 993;
  return new ImapFlow({
    host, port, secure: true,
    auth: { user: config.email, pass: config.password },
    logger: false,
    connectionTimeout: FETCH_TIMEOUT_MS,
    greetingTimeout:   FETCH_TIMEOUT_MS,
    socketTimeout:     FETCH_TIMEOUT_MS,
  });
}

function friendlyError(err: unknown, host: string, port: number): Error {
  const msg = (err as Error).message ?? String(err);
  if (/AUTHENTICATIONFAILED|Invalid credentials|Authentication failed/i.test(msg))
    return new Error('Authentication failed. For Gmail, use an App Password (myaccount.google.com/apppasswords).');
  if (/ETIMEDOUT|ECONNREFUSED|ENOTFOUND/i.test(msg))
    return new Error(`Cannot connect to IMAP server (${host}:${port}). Check your host settings.`);
  return err as Error;
}

/**
 * Check inbox for a payment notification matching `amount` from `provider`.
 * Returns the message UID string on match, or null if not found.
 * Pass `alreadyUsedUids` to prevent the same email confirming two orders.
 */
export async function checkEmailPayment(
  provider: PaymentProvider,
  amount: number,
  orderId: string,
  config: PaymentEmailConfig,
  lookbackMinutes = 45,
  alreadyUsedUids: Set<string> = new Set(),
): Promise<string | null> {
  const host = config.host || inferHost(config.email);
  const port = config.port || 993;
  const client = makeClient(config);

  try {
    await client.connect();
    const lock = await client.getMailboxLock('INBOX');

    try {
      const since = new Date(Date.now() - lookbackMinutes * 60 * 1000);
      const uids  = await client.search({ since }, { uid: true }) as unknown as number[];
      if (!uids || uids.length === 0) return null;

      for await (const msg of client.fetch([...uids].reverse(), { uid: true, envelope: true, source: true }, { uid: true }) as AsyncIterable<any>) {
        const uid     = String(msg.uid);
        if (alreadyUsedUids.has(uid)) continue;

        const subject = msg.envelope?.subject ?? '';
        const from    = msg.envelope?.from?.[0]?.address ?? '';
        const date    = msg.envelope?.date;

        if (!isProviderEmail(provider, subject, from)) continue;

        // Time gate
        if (date) {
          const cutoff = Date.now() - lookbackMinutes * 60 * 1000;
          if (new Date(date).getTime() < cutoff) continue;
        }

        const rawText   = (msg.source ?? Buffer.alloc(0)).toString('utf8');
        const plainText = extractPlainText(rawText);
        const combined  = (subject + ' ' + plainText).toLowerCase();

        const amountMatches = extractAmounts(combined).some(a => Math.abs(a - amount) < 0.015);

        if (provider === 'cashapp') {
          // Cash App: require strict sender, amount match, and memo/body contains orderId
          const strictSender = /^cash@square\.com$/i.test(from) || /cash@square\.com/i.test(from);
          if (!strictSender) continue;
          if (!amountMatches) continue;
          const memo = extractMemo(plainText);
          const memoHasOrderId = orderId
            ? (memo.toLowerCase().includes(orderId.toLowerCase()) || combined.includes(orderId.toLowerCase()))
            : true;
          if (memoHasOrderId) return uid;
          continue;
        }

        // All other providers: require amount match AND (orderId found OR no orderId given).
        // Requiring both prevents false positives when multiple payments arrive for different orders.
        const hasOrderId = orderId ? combined.includes(orderId.toLowerCase()) : true;
        if (amountMatches && hasOrderId) return uid;
      }

      return null;
    } finally {
      lock.release();
    }
  } catch (err) {
    throw friendlyError(err, host, port);
  } finally {
    try { client.close(); } catch {}
  }
}

export interface PaymentEmailInspectResult {
  uid: string;
  from: string;
  subject: string;
  date: string;
  provider: PaymentProvider | 'unknown';
  amountsFound: number[];
  memo: string;
  bodySnippet: string;
  matched: boolean;
  matchReason: string;
}

/**
 * Scan the inbox and return detailed info about the most recent payment-looking email.
 * If provider/amount are given, also reports whether it would match.
 */
export async function inspectLatestPaymentEmail(
  config: PaymentEmailConfig,
  provider?: PaymentProvider,
  amount?: number,
  lookbackMinutes = 45,
): Promise<PaymentEmailInspectResult | null> {
  const host = config.host || inferHost(config.email);
  const port = config.port || 993;
  const client = makeClient(config);

  try {
    await client.connect();
    const lock = await client.getMailboxLock('INBOX');

    try {
      const since = new Date(Date.now() - lookbackMinutes * 60 * 1000);
      const uids = await client.search({ since }, { uid: true }) as unknown as number[];
      if (!uids || uids.length === 0) return null;

      const allProviders: PaymentProvider[] = ['cashapp', 'venmo', 'zelle', 'paypal'];

      for await (const msg of client.fetch([...uids].reverse(), { uid: true, envelope: true, source: true }, { uid: true }) as AsyncIterable<any>) {
        const subject   = msg.envelope?.subject ?? '';
        const from      = msg.envelope?.from?.[0]?.address ?? '';
        const date      = msg.envelope?.date ? new Date(msg.envelope.date).toISOString() : 'unknown';
        const rawText   = (msg.source ?? Buffer.alloc(0)).toString('utf8');
        const plainText = extractPlainText(rawText);
        const combined  = (subject + ' ' + plainText).toLowerCase();
        const amountsFound = extractAmounts(combined);
        const memo      = extractMemo(plainText);

        // For inspect: match Cash App by sender only (so we can see what's in the inbox and report the subject).
        // For all other providers, use the full isProviderEmail check.
        const senderOnlyMatch = (p: PaymentProvider) => {
          const pat = PROVIDER_PATTERNS[p];
          return p === 'cashapp'
            ? pat.from.some(r => r.test(from))
            : pat.from.some(r => r.test(from)) || pat.subject.some(r => r.test(subject));
        };
        const detectedProvider = provider
          ? (senderOnlyMatch(provider) ? provider : undefined)
          : allProviders.find(p => senderOnlyMatch(p));

        if (!detectedProvider) continue;

        const resolvedProvider: PaymentProvider | 'unknown' = detectedProvider ?? 'unknown';

        let matched = false;
        let matchReason = 'No match criteria given';

        if (provider && amount !== undefined) {
          const amountMatch   = amountsFound.some(a => Math.abs(a - amount) < 0.015);
          const providerMatch = isProviderEmail(provider, subject, from);
          matched = amountMatch && providerMatch;
          if (matched) {
            matchReason = `Provider matched + amount $${amount} found`;
          } else if (amountMatch) {
            matchReason = `Amount $${amount} found but subject did not match payment patterns (subject: "${subject}")`;
          } else if (providerMatch) {
            matchReason = `Provider matched but amount $${amount} not found (saw: ${amountsFound.map(a => '$' + a).join(', ') || 'none'})`;
          } else {
            matchReason = `Subject did not match payment patterns (subject: "${subject}") and amount $${amount} not found`;
          }
        }

        // First 400 chars of decoded plain text for debugging
        const bodySnippet = plainText.replace(/\s+/g, ' ').trim().slice(0, 400);
        return { uid: String(msg.uid), from, subject, date, provider: resolvedProvider, amountsFound, memo, bodySnippet, matched, matchReason };
      }

      return null;
    } finally {
      lock.release();
    }
  } catch (err) {
    throw friendlyError(err, host, port);
  } finally {
    try { client.close(); } catch {}
  }
}

/**
 * Verify IMAP credentials without searching for any payment.
 * Throws a descriptive error on failure.
 */
export async function testPaymentEmail(config: PaymentEmailConfig): Promise<true> {
  const host = config.host || inferHost(config.email);
  const port = config.port || 993;
  const client = makeClient(config);
  try {
    await client.connect();
    const lock = await client.getMailboxLock('INBOX');
    lock.release();
    return true;
  } catch (err) {
    throw friendlyError(err, host, port);
  } finally {
    try { client.close(); } catch {}
  }
}

File: email_payments.py
"""
email_payments.py — IMAP-based payment verification.
Supports Cash App, Venmo, Zelle, and PayPal notification emails.
Mirrors email-payments.ts exactly.
"""
import re
import html
from datetime import datetime, timezone, timedelta
from typing import Optional, Literal, TypedDict
from dataclasses import dataclass, field

from imapclient import IMAPClient

PaymentProvider = Literal["cashapp", "venmo", "zelle", "paypal"]

FETCH_TIMEOUT_S = 15

KNOWN_HOSTS: dict[str, str] = {
    "gmail.com":      "imap.gmail.com",
    "googlemail.com": "imap.gmail.com",
    "outlook.com":    "outlook.office365.com",
    "hotmail.com":    "outlook.office365.com",
    "live.com":       "outlook.office365.com",
    "yahoo.com":      "imap.mail.yahoo.com",
    "ymail.com":      "imap.mail.yahoo.com",
    "icloud.com":     "imap.mail.me.com",
    "me.com":         "imap.mail.me.com",
}


@dataclass
class PaymentEmailConfig:
    email: str
    password: str
    host: Optional[str] = None
    port: int = 993


def _infer_host(email: str) -> str:
    domain = email.split("@")[-1].lower() if "@" in email else ""
    return KNOWN_HOSTS.get(domain, "imap.gmail.com")


PROVIDER_PATTERNS: dict[str, dict] = {
    "cashapp": {
        "from": [re.compile(r"cash@square\.com", re.I), re.compile(r"no-?reply@cash\.app", re.I), re.compile(r"@cashapp\.com", re.I), re.compile(r"cash@cashapp\.com", re.I)],
        "subject": [re.compile(r"sent you \$", re.I), re.compile(r"you received \$", re.I), re.compile(r"cash app.*payment", re.I), re.compile(r"payment.*cash app", re.I), re.compile(r"you've got cash", re.I)],
    },
    "venmo": {
        "from": [re.compile(r"venmo@venmo\.com", re.I), re.compile(r"no-?reply@venmo\.com", re.I)],
        "subject": [re.compile(r"paid you", re.I), re.compile(r"sent you \$", re.I), re.compile(r"you've got money", re.I), re.compile(r"payment.*received", re.I), re.compile(r"venmo.*payment", re.I)],
    },
    "zelle": {
        "from": [re.compile(r"zelle", re.I), re.compile(r"noreply@chase\.com", re.I), re.compile(r"alerts@bankofamerica\.com", re.I), re.compile(r"alerts@.*wellsfargo\.com", re.I), re.compile(r"alerts@.*usbank\.com", re.I), re.compile(r"alerts@.*capitalone\.com", re.I), re.compile(r"notify@.*zellepay\.com", re.I)],
        "subject": [re.compile(r"you received.*zelle", re.I), re.compile(r"received money with zelle", re.I), re.compile(r"zelle.*payment received", re.I), re.compile(r"payment.*received.*zelle", re.I), re.compile(r"you have received", re.I), re.compile(r"zelle.*you.*received", re.I)],
    },
    "paypal": {
        "from": [re.compile(r"service@paypal\.com", re.I), re.compile(r"service@intl\.paypal\.com", re.I), re.compile(r"paypal@e\.paypal\.com", re.I), re.compile(r"noreply@paypal\.com", re.I)],
        "subject": [re.compile(r"you've got money", re.I), re.compile(r"you received \$", re.I), re.compile(r"payment.*received", re.I), re.compile(r"sent you \$", re.I), re.compile(r"paypal.*payment", re.I), re.compile(r"money.*received", re.I)],
    },
}


def _is_provider_email(provider: str, subject: str, from_addr: str) -> bool:
    p = PROVIDER_PATTERNS[provider]
    from_match = any(r.search(from_addr) for r in p["from"])
    subject_match = any(r.search(subject) for r in p["subject"])
    if provider == "cashapp":
        return from_match and subject_match
    return from_match or subject_match


def _extract_memo(text: str) -> str:
    patterns = [
        re.compile(r"\bFor:?\s+([A-Z0-9][A-Z0-9\-_]{1,40})\b", re.I),
        re.compile(r"\bFor\s*[\r\n]+\s*([A-Z0-9][A-Z0-9\-_]{1,40})\b", re.I),
        re.compile(r"\bNote:\s*([^\n\r<]{2,100})", re.I),
        re.compile(r"\bNote\s*[\r\n]+\s*([^\n\r<]{2,100})", re.I),
        re.compile(r"\bMemo:\s*([^\n\r<]{2,100})", re.I),
        re.compile(r"\bMessage:\s*([^\n\r<]{2,100})", re.I),
        re.compile(r"\bDescription:\s*([^\n\r<]{2,100})", re.I),
        re.compile(r"\bReference:\s*([^\n\r<]{2,100})", re.I),
        re.compile(r'note:?\s+"([^"]{2,100})"', re.I),
        re.compile(r"note:?\s+'([^']{2,100})'", re.I),
    ]
    for pattern in patterns:
        m = pattern.search(text)
        if not m:
            continue
        val = " ".join(m.group(1).split()).strip()
        if len(val) < 2:
            continue
        if re.search(r"https?://", val, re.I):
            continue
        if re.search(r"version=|cipher|bits\d|sha\d{3}", val, re.I):
            continue
        if re.search(r"[0-9a-f]{20,}", val, re.I):
            continue
        if re.search(r"@\w+\.\w+", val):
            continue
        if re.match(r"^(more information|questions|your account|any questions|details|security|help|click|view|manage|update|visit)\b", val, re.I):
            continue
        return val
    return ""


def _strip_rfc822_headers(raw: str) -> str:
    idx = raw.find("\r\n\r\n")
    if idx != -1:
        return raw[idx + 4:]
    idx = raw.find("\n\n")
    if idx != -1:
        return raw[idx + 2:]
    return raw


def _decode_quoted_printable(s: str) -> str:
    s = re.sub(r"=\r?\n", "", s)
    s = re.sub(r"=([0-9A-Fa-f]{2})", lambda m: chr(int(m.group(1), 16)), s)
    return s


def _decode_html_entities(s: str) -> str:
    return (
        s.replace("&amp;", "&")
        .replace("&nbsp;", " ")
        .replace("&#160;", " ")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", '"')
        .replace("&#39;", "'")
        .replace("&#x27;", "'")
        .replace("&#x2F;", "/")
    )
    # Also handle numeric HTML entities
    result = re.sub(r"&#(\d+);", lambda m: chr(int(m.group(1))), s)
    return result


def _extract_plain_text(raw: str) -> str:
    body = _strip_rfc822_headers(raw)
    decoded = _decode_quoted_printable(body)
    no_tags = re.sub(r"<[^>]+>", " ", decoded)
    return _decode_html_entities(no_tags)


def _extract_amounts(text: str) -> list[float]:
    amounts = []
    for m in re.finditer(r"[+\-]?\$\s*([\d]{1,6}(?:,\d{3})*(?:\.\d{1,2})?)", text, re.I):
        try:
            v = float(m.group(1).replace(",", ""))
            if not (0 < v < 10000):
                continue
            amounts.append(v)
        except ValueError:
            pass
    return list(dict.fromkeys(amounts))  # deduplicate preserving order


def _friendly_error(err: Exception, host: str, port: int) -> Exception:
    msg = str(err)
    if re.search(r"AUTHENTICATIONFAILED|Invalid credentials|Authentication failed", msg, re.I):
        return ValueError("Authentication failed. For Gmail, use an App Password (myaccount.google.com/apppasswords).")
    if re.search(r"ETIMEDOUT|ECONNREFUSED|ENOTFOUND|timed out|refused|Name or service not known", msg, re.I):
        return ConnectionError(f"Cannot connect to IMAP server ({host}:{port}). Check your host settings.")
    return err


def _make_client(config: PaymentEmailConfig) -> IMAPClient:
    host = config.host or _infer_host(config.email)
    return IMAPClient(host, port=config.port, ssl=True, timeout=FETCH_TIMEOUT_S)


async def check_email_payment(
    provider: str,
    amount: float,
    order_id: str,
    config: PaymentEmailConfig,
    lookback_minutes: int = 45,
    already_used_uids: Optional[set] = None,
) -> Optional[str]:
    """
    Returns UID string if a matching payment email is found, else None.
    """
    if already_used_uids is None:
        already_used_uids = set()

    host = config.host or _infer_host(config.email)
    port = config.port

    import asyncio
    loop = asyncio.get_event_loop()

    def _sync_check():
        client = _make_client(config)
        try:
            client.login(config.email, config.password)
            client.select_folder("INBOX")
            since = datetime.now(timezone.utc) - timedelta(minutes=lookback_minutes)
            uids = client.search(["SINCE", since.date()])
            if not uids:
                return None

            for uid in reversed(uids):
                uid_str = str(uid)
                if uid_str in already_used_uids:
                    continue
                msgs = client.fetch([uid], ["ENVELOPE", "RFC822"])
                if uid not in msgs:
                    continue
                data = msgs[uid]
                envelope = data.get(b"ENVELOPE")
                raw_bytes = data.get(b"RFC822", b"")

                subject = ""
                from_addr = ""
                msg_date = None

                if envelope:
                    if envelope.subject:
                        try:
                            subj_raw = envelope.subject
                            if isinstance(subj_raw, bytes):
                                subject = subj_raw.decode("utf-8", errors="replace")
                            else:
                                subject = str(subj_raw)
                        except Exception:
                            subject = ""
                    if envelope.from_:
                        f = envelope.from_[0]
                        mb = (f.mailbox or b"").decode("utf-8", errors="replace") if isinstance(f.mailbox, bytes) else str(f.mailbox or "")
                        host_part = (f.host or b"").decode("utf-8", errors="replace") if isinstance(f.host, bytes) else str(f.host or "")
                        from_addr = f"{mb}@{host_part}"
                    if envelope.date:
                        msg_date = envelope.date

                if not _is_provider_email(provider, subject, from_addr):
                    continue

                if msg_date:
                    cutoff = datetime.now(timezone.utc) - timedelta(minutes=lookback_minutes)
                    if isinstance(msg_date, datetime):
                        dt = msg_date if msg_date.tzinfo else msg_date.replace(tzinfo=timezone.utc)
                        if dt < cutoff:
                            continue

                raw_text = raw_bytes.decode("utf-8", errors="replace") if isinstance(raw_bytes, bytes) else str(raw_bytes)
                plain_text = _extract_plain_text(raw_text)
                combined = (subject + " " + plain_text).lower()

                amount_matches = any(abs(a - amount) < 0.015 for a in _extract_amounts(combined))

                if provider == "cashapp":
                    strict_sender = bool(re.search(r"cash@square\.com", from_addr, re.I))
                    if not strict_sender:
                        continue
                    if not amount_matches:
                        continue
                    memo = _extract_memo(plain_text)
                    has_order_id = (
                        order_id.lower() in memo.lower() or order_id.lower() in combined
                    ) if order_id else True
                    if has_order_id:
                        return uid_str
                    continue

                has_order_id = order_id.lower() in combined if order_id else True
                if amount_matches and has_order_id:
                    return uid_str

            return None
        except Exception as e:
            raise _friendly_error(e, host, port)
        finally:
            try:
                client.logout()
            except Exception:
                pass

    return await loop.run_in_executor(None, _sync_check)


async def inspect_latest_payment_email(
    config: PaymentEmailConfig,
    provider: Optional[str] = None,
    amount: Optional[float] = None,
    lookback_minutes: int = 45,
) -> Optional[dict]:
    host = config.host or _infer_host(config.email)
    port = config.port
    import asyncio
    loop = asyncio.get_event_loop()

    def _sync_inspect():
        client = _make_client(config)
        try:
            client.login(config.email, config.password)
            client.select_folder("INBOX")
            since = datetime.now(timezone.utc) - timedelta(minutes=lookback_minutes)
            uids = client.search(["SINCE", since.date()])
            if not uids:
                return None

            all_providers = ["cashapp", "venmo", "zelle", "paypal"]

            for uid in reversed(uids):
                msgs = client.fetch([uid], ["ENVELOPE", "RFC822"])
                if uid not in msgs:
                    continue
                data = msgs[uid]
                envelope = data.get(b"ENVELOPE")
                raw_bytes = data.get(b"RFC822", b"")

                subject = ""
                from_addr = ""
                msg_date = "unknown"

                if envelope:
                    if envelope.subject:
                        try:
                            sb = envelope.subject
                            subject = sb.decode("utf-8", errors="replace") if isinstance(sb, bytes) else str(sb)
                        except Exception:
                            pass
                    if envelope.from_:
                        f = envelope.from_[0]
                        mb = (f.mailbox or b"").decode("utf-8", errors="replace") if isinstance(f.mailbox, bytes) else str(f.mailbox or "")
                        hp = (f.host or b"").decode("utf-8", errors="replace") if isinstance(f.host, bytes) else str(f.host or "")
                        from_addr = f"{mb}@{hp}"
                    if envelope.date:
                        dt = envelope.date
                        if isinstance(dt, datetime):
                            msg_date = dt.isoformat()

                def sender_only_match(p: str) -> bool:
                    pats = PROVIDER_PATTERNS[p]
                    if p == "cashapp":
                        return any(r.search(from_addr) for r in pats["from"])
                    return any(r.search(from_addr) for r in pats["from"]) or any(r.search(subject) for r in pats["subject"])

                if provider:
                    detected = provider if sender_only_match(provider) else None
                else:
                    detected = next((p for p in all_providers if sender_only_match(p)), None)

                if not detected:
                    continue

                raw_text = raw_bytes.decode("utf-8", errors="replace") if isinstance(raw_bytes, bytes) else str(raw_bytes)
                plain_text = _extract_plain_text(raw_text)
                combined = (subject + " " + plain_text).lower()
                amounts_found = _extract_amounts(combined)
                memo = _extract_memo(plain_text)

                matched = False
                match_reason = "No match criteria given"
                if provider and amount is not None:
                    amount_match = any(abs(a - amount) < 0.015 for a in amounts_found)
                    provider_match = _is_provider_email(provider, subject, from_addr)
                    matched = amount_match and provider_match
                    if matched:
                        match_reason = f"Provider matched + amount ${amount} found"
                    elif amount_match:
                        match_reason = f"Amount ${amount} found but subject did not match payment patterns (subject: \"{subject}\")"
                    elif provider_match:
                        match_reason = f"Provider matched but amount ${amount} not found (saw: {', '.join('$'+str(a) for a in amounts_found) or 'none'})"
                    else:
                        match_reason = f"Subject did not match and amount ${amount} not found"

                body_snippet = " ".join(plain_text.split())[:400]
                return {
                    "uid": str(uid),
                    "from": from_addr,
                    "subject": subject,
                    "date": msg_date,
                    "provider": detected,
                    "amountsFound": amounts_found,
                    "memo": memo,
                    "bodySnippet": body_snippet,
                    "matched": matched,
                    "matchReason": match_reason,
                }
            return None
        except Exception as e:
            raise _friendly_error(e, host, port)
        finally:
            try:
                client.logout()
            except Exception:
                pass

    import asyncio
    loop = asyncio.get_event_loop()
    return await loop.run_in_executor(None, _sync_inspect)


async def test_payment_email(config: PaymentEmailConfig) -> bool:
    host = config.host or _infer_host(config.email)
    port = config.port
    import asyncio
    loop = asyncio.get_event_loop()

    def _sync_test():
        client = _make_client(config)
        try:
            client.login(config.email, config.password)
            client.select_folder("INBOX")
            return True
        except Exception as e:
            raise _friendly_error(e, host, port)
        finally:
            try:
                client.logout()
            except Exception:
                pass

    return await loop.run_in_executor(None, _sync_test)


# Zelle shim — same as TypeScript's zelle.ts re-export
check_zelle_payment = check_email_payment
test_zelle_connection = test_payment_email
